[CORE-8530] Integrate struct evolution machinery into catalog_schema_manager

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR enables full-featured struct evolution into the iceberg record translation path.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Enables full struct schema evolution for Iceberg topics

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
ice/compat: Fix for_each_field vector usage

For structs, prime the stack with all top-level fields instead of visiting
one by one. This results in fewer allocations.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ice/datatypes: Improve evolution metadata API for nested_field

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dt/dl: Integration testing for schema compat integration

test that:
- table schema updates work
- valid evolved schemas are backwards compatible w/ historical rows
- adding columns never results in reading pre-existing values from another column
- removing a column and adding it back in a subsequent update creates a NEW
  column that does not collide with the old 'version'
- attempting to read from a dropped column fails gracefully
- changing the order of columns doesn't change the values associated with the
  a column or field name

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ice/compat: Introduce try_fill_field_ids

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ice/schema: Make assign_fresh_ids respect evolution metadata

Previously we would assign an ID to every field in the input schema starting
from the minimum valid ID (1).

Instead, we assign IDs _only_ to those fields which did not have one assigned
during compat checks, and (instead of starting from 'id_t{1}') starting from some
ID specified at the interface. In the case of applying a schema update, this
should be one past the last_column_id recorded in table metadata.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ice/update_schema: Allow updates with max ID < meta.last_column_id

In effect, this allows us to support full field removal.

We still enforce that meta.last increases monotonically. This invariant
is important to avoid reusing column IDs, which would in turn violate
specified schema evolution correctness guarantees.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dl/catalog_schema_manager: Wire up compatibility checks

Introduce check_schema_compat, called on the ensure_schema path, to call into
the compatibility layer. Returns an error, or an indication that the schema
changed, or nothing (no schema change).

Also update fill_field_ids to use the compat machinery to support filling field
IDs in the presence of type changes, changes in field order, or records with
missing fields. Errors iff not all destination fields could be filled.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61083-->
<!--227dd0fb-73ee-4463-b26e-6f6fe82fd698-->
<!--6cf1b5b3-e8bf-40f5-84a8-ba7c6db2c614-->
<!--2c2d953c-7e6c-4ff4-bb30-7f92e4c2667a-->
<!--title-->
## Retry command for [Build#61083](https://buildkite.com/redpanda/redpanda/builds/61083)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionSmokeTest.test_evolving_avro_schemas@{"cloud_storage_type":1,"query_engine":"trino"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionSmokeTest.test_evolving_avro_schemas@{"cloud_storage_type":1,"query_engine":"spark"}
```
<!--/body-->

oleiman commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--afb231aa-06fb-4d58-8f46-7fbe8483492c-->
<!--TEST RESULTS-->
<!--33bc77b5-9851-4542-b7ef-aa5422d4d71f-->
<!--TEST RESULTS-->
<!--de928378-4180-4eac-8564-71fb9f78a985-->
<!--TEST RESULTS-->
<!--bcb0a039-c2a0-48a3-91cb-9d2bac809d60-->
<!--TEST RESULTS-->
<!--05264f30-dc5d-4ad7-8cfb-1d1503294f77-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61091">build#61091</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61091#0194922f-073a-4947-b34b-3353530985e5 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61091#01949241-b80a-4f89-a11f-3a319048aef9 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61182">build#61182</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.availability_test.AvailabilityTests.test_availability_when_one_node_failed | ducktape | https://buildkite.com/redpanda/redpanda/builds/61182#01949a86-78ad-47df-9823-f07155ce8822 | FLAKY | 1/2 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61182#01949a86-78ae-4205-b067-94d41afa8aec | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61281">build#61281</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61281#0194abcc-3f7d-473e-9a6a-29388045798e | FLAKY | 1/3 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_creating_and_listing_migrations | ducktape | https://buildkite.com/redpanda/redpanda/builds/61281#0194abcc-3f7c-46b5-a249-3bdbe10ff19a | FLAKY | 1/2 |
| rptest.tests.partition_force_reconfiguration_test.PartitionForceReconfigurationTest.test_basic_reconfiguration.acks=-1.restart=False.controller_snapshots=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61281#0194abc6-d457-4756-bb37-b6e4f440901b | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_cross_shard.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61281#0194abcc-3f7e-46b4-80c7-0367aaa74002 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61320">build#61320</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61320#0194afea-9668-4506-838b-04afc600755d | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61320#0194b005-9b89-4463-ad7e-3f6f78c4604c | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_higher_level_migration_api | ducktape | https://buildkite.com/redpanda/redpanda/builds/61320#0194afea-9665-4f04-bfb0-1905a9a5f6b2 | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61320#0194b005-9b89-4463-ad7e-3f6f78c4604c | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61405">build#61405</a></summary>


| test_id                                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61405#0194b8c9-f043-4cdf-b85b-df86ce66afde | FLAKY         | 1/2      |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_basic.cloud_storage_type=CloudStorageType.S3.filesystem_catalog_mode=False                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61405#0194b8c5-0994-4915-a786-bb9b1643288a | FLAKY         | 1/2      |
| rptest.tests.datalake.datalake_dlq_test.DatalakeDLQTest.test_invalid_record_action_runtime_change.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.use_topic_property=True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61405#0194b8c9-f044-4800-a9ac-ec7da0834308 | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61405#0194b8c9-f044-4315-903f-869b22e9384f | FLAKY         | 1/3      |
</details>
<!--/body-->

oleiman commented:
/cdt
tests/rptest/tests/datalake/schema_evolution_test.py

oleiman commented on src/v/iceberg/compatibility_utils.h:
```diff
@@ -163,4 +162,7 @@ schema_errc_result for_each_field(
       s, fn, [](const nested_field*) { return true; });
 }
 
+// TODO(oren): docstring
```
missed a docstring

oleiman committed:
dl/record_translator: Force all nested fields to non-required

nullability semantics differ significantly between schema definition languages.
without loss of generality, we can internally treat all fields as nullable to
avoid complexity associated with maintaining requiredness invariants across
schema versions.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/cdt
tests/rptest/tests/datalake/schema_evolution_test.py
dt-repeat=5

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/20ddb11cee964550a036964e0fc4e09824e48b1d..598341080a307fb3c3c6cd9d33779ab256a59530) moved "historical writes" commits into a separate PR #24955

rockwotj commented on src/v/iceberg/datatypes_json.cc:
```diff
@@ -177,6 +177,10 @@ void rjson_serialize(iceberg::json_writer& w, const iceberg::nested_field& f) {
     w.Bool(bool(f.required));
     w.Key("type");
     rjson_serialize(w, f.type);
+    w.Key("initial-default");
+    w.Null();
+    w.Key("write-default");
+    w.Null();
```
Wait we make everything optional? Which is why this is valid? Should we only include these if `f.required`?

rockwotj replied:
Oh I see it's a later commit.... Should we reorder? :)

oleiman replied:
yeah fair point.

oleiman replied:
done

rockwotj commented on src/v/datalake/record_translator.cc:
```diff
@@ -188,6 +189,11 @@ structured_data_translator::build_type(std::optional<resolved_type> val_type) {
     if (val_type.has_value()) {
         val_id = std::move(val_type->id);
         auto& struct_type = std::get<iceberg::struct_type>(val_type->type);
+        // TODO: could make this configurable at some point
+        std::ignore = iceberg::for_each_field(
+          struct_type, [](iceberg::nested_field* f) {
+              f->required = iceberg::field_required::no;
+          });
```
> nested fields

This is essentially all fields right? Because the type root is a struct type correct?

oleiman replied:
yup

rockwotj reviewed (commented):
Seems fairly straightforward, I had a comment on the commit order throwing me off. I haven't gone over the last two commits yet, I promise I will do so tomorrow.

oleiman commented:
> Seems fairly straightforward,

Ya. Feels a little "too straightforward", but maybe it just is.

> I haven't gone over the last two commits yet, I promise I will do so tomorrow.

Cool thanks. The last two commits in this PR are the meat of the change & new tests, but do note that I spun off part of the stack into a separate PR for handling incoming records from an "old" schema here: #24955 - that bit might be worthy of discussion w.r.t. product def'n.

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/598341080a307fb3c3c6cd9d33779ab256a59530..7d81009699d72ef16cfb90f893160c4bc5c4245f) reorder commits

andrwng commented on src/v/iceberg/schema.cc:
```diff
@@ -105,9 +106,14 @@ void schema::assign_fresh_ids() {
+        } else if (field->is_add()) {
+            field->id = next_id++;
         }
+        vassert(
+          field->id != 0,
+          "Nested field '{}' wasn't assigned an ID",
+          field->name);
```
This feels like it should always be true, but is there a guarantee in code that this never fires? Could a buggy catalog cause a crash, for example?

oleiman replied:
yeah, i've gone back and forth on this. I think the assertion holds for structs processed through record translation, but...

>could a buggy catalog cause a crash

good point, hadn't considered that. I think it's fair to say that a field w/ a zero ID here is strictly ill formed, but since that value could have come from disk probably a non-fatal error would be more appropriate.

andrwng commented on src/v/iceberg/table_update_applier.cc:
```diff
@@ -34,16 +34,13 @@ struct update_applying_visitor {
-                return outcome::unexpected_state;
-            }
-            meta.last_column_id = new_last_col_id;
+            // NOTE: we can drop fields now, so new_last_col_id could be
```
nit: drop the "now" to make this comment context/history-agnostic

oleiman replied:
done

andrwng commented on src/v/iceberg/tests/update_schema_action_test.cc:
```diff
@@ -183,21 +183,25 @@ TEST_F(UpdateSchemaActionTest, TestNewMultipleSchemas) {
-    res = tx.set_schema(std::move(new_new_schema)).get();
-    ASSERT_TRUE(res.has_error());
-    ASSERT_TRUE(tx.error().has_value());
+    ASSERT_FALSE(res.has_error());
+    ASSERT_FALSE(tx.error().has_value());
```
Can we assert that the last column id is what we expect?

oleiman replied:
ya good catch

oleiman replied:
done

andrwng commented on src/v/datalake/record_translator.cc:
```diff
@@ -188,6 +189,11 @@ structured_data_translator::build_type(std::optional<resolved_type> val_type) {
         val_id = std::move(val_type->id);
         auto& struct_type = std::get<iceberg::struct_type>(val_type->type);
+        // TODO: could make this configurable at some point
+        std::ignore = iceberg::for_each_field(
+          struct_type, [](iceberg::nested_field* f) {
+              f->required = iceberg::field_required::no;
+          });
```
>without loss of generality, we can internally treat all fields as nullable to
>avoid complexity associated with maintaining requiredness invariants across
>schema versions.

I'm not sure I follow this -- is this really _just_ an internal detail? Doesn't it get propagated to users in their tables?

andrwng replied:
If the intention here is really to work around how strict Avro is, I wonder if it's less surprising to either:
- not allow evolving Avro schemas
- require evolution of Avro schemas to be in the form of adding null-unions

I personally don't feel strongly that either is correct, but I just found this part of the PR surprising (sorry if we decided this a while ago and I'm just forgetting) and want to make sure this is the least surprising option for users

oleiman replied:
> is this really just an internal detail

No not really, just poor phrasing. What I mean is "strictly within iceberg land". Trying to emphasize that the kafka record serde language imposes its own compatibility invariants, so on the Iceberg side we can, in principle, stay in the business of aggressively maintaining _table_ compatibility, with the explicit assumption that record compat semantics are handled elsewhere.

> if the intention here is really to work around how strict Avro is

Sort of. I think we're still working around an impedance mismatch between Avro or protobuf (or JSON as the case may be) compat vs Iceberg compat.

> require evolution of Avro schemas to be in the form of adding null-unions

is a good example. For Avro serde purposes `{"name": "foo", "type": ["null", "int"]}` w/ a null default is effectively an optional field called "foo". BUT in Iceberg schema we translate that into a *required* field "foo" of type `struct<union_opt_1:int>` where `union_opt_1` is non-required.

We could do something like make the null-union structs optional during type resolution, but it still feels like we're smashing abstractions together in an uncomfortable way.

> not allow evolving Avro schemas

Right. My read is that we should either be maximally restrictive or maximally permissive to avoid reasoning about (and documenting) these special cases. 

My intuitive preference is to start with the permissive version because it feels more "capable" in a product sense. My argument to "generality" is that we're translating records that are already subject to an SR schema (w/ associated compat rules), so the schema language's requiredness invariants have _already been applied_ prior to translation. So even if the columns are "nullable" for compat purposes, the _values_ stored there are still subject to the semantics of the corresponding record stream.

> sorry if we decided this a while ago and I'm just forgetting

not at all. this came up during implementation and I think now is the time to settle it. might be easier to chat through face to face (and maybe pull in @mattschumpert and anyone else interested).

oleiman replied:
Ah, other thing i wanted to mention is that the databricks connector does this (configurable) for the same(ish) reason, which is where i got the idea. So not without precedent 🤷‍♂️ 

Config is `iceberg.tables.schema-force-optional` per https://github.com/databricks/iceberg-kafka-connect#configuration

andrwng replied:
Thanks for the details here!

>but it still feels like we're smashing abstractions together in an uncomfortable way.

Yea generally I agree that mixing Iceberg and Avro/Kafka semantics feels pretty ick. At this point in the data path, it feels cleaner to think of the question, "what should my Iceberg table do?" without thinking about Kafka schema semantics at all.

>My intuitive preference is to start with the permissive version because it feels more "capable" in a product sense.

Ah yea that's a fair point, though there's a case to be made that it may be less problematic to do the less surprising, more limited thing first, and then extend our capabilities later with some additional config.

>so the schema language's requiredness invariants have already been applied prior to translation

Yea this is also fair. The remaining question though is whether requiredness of Kafka schema holds meaning to query engine users.

I guess another option to consider: what about only allowing adding required fields that have defaults? I seem to recall some work around serializing values moving in that direction anyway. Did it turn out to be not viable?

Let's set up some time to chat about this -- it'd be good to get product's thoughts around users' table expectations and default behaviors.

oleiman replied:
Ya, I'll try to set something up for tomorrow.

> whether requiredness of Kafka schema holds meaning to query engine users

 Yeah, it's an open question from my POV. I imagine there are distinct personas here - e.g. 
- "analyst Iceberg end user for whom table schema is the source of truth" 
- "infra engineer who may directly control both the source (topic) and sink (table)".

> do the less surprising thing

100% agree that should be the goal, and I think we get to draw the map here (within reason) since the ecosystem is evolving so quickly. But yeah, mostly a product question at this point.

> what about only allowing adding required fields that have defaults?

That should be the north star I think [1]. The tricky bit is how to specify defaults - do they come from the source schema? Or maybe we do what protobuf does (i.e. a string's default is always the empty string, default bool is false, etc)? I think we talked about this last week and decided to punt for GA, but the value serialization code should be fit for purpose.

[1] Also note that the spec requires default values for any non-optional field added to an existing schema.

rockwotj replied:
> Or maybe we do what protobuf does (i.e. a string's default is always the empty string, default bool is false, etc)?

Just FYI it's possible to override the default value in protobuf. We should be respecting that.

TBH, I'm a fan of making everything nullable (not required). It looks like the kafka connector has an option for this: `iceberg.tables.schema-force-optional`, FWIW Snowflake's connector seems to always add new columns as null and not mess with defaults.

oleiman replied:
Don't think we reached as clear consensus as I would have hoped, but that's ok. My read is that the permissive approach should be sufficient (in a product sense) for initial GA, modulo any lingering concerns from this thread.

I expect we'll be asked to support a more strict adherence to record schema before long, with requiredness and defaults and whatnot, but I don't think we're painting ourselves into a corner w/ the proposed implementation. Happy to hear differently from anyone though.

rockwotj replied:
Agreed. I think from @mattschumpert the thought was that this is a good default, which is the more important thing. We can also ad options to tune this behavior for customers later on.

andrwng commented on src/v/iceberg/datatypes_json.cc:
```diff
@@ -177,6 +177,10 @@ void rjson_serialize(iceberg::json_writer& w, const iceberg::nested_field& f) {
     w.Bool(bool(f.required));
     w.Key("type");
     rjson_serialize(w, f.type);
+    w.Key("initial-default");
+    w.Null();
+    w.Key("write-default");
```
Maybe leave a TODO explaining that this isn't a bug to always serialize null, and that we just don't support required fields yet?

Though the `redpanda` struct has some required fields -- is it invalid to serialize nulls as their defaults?

oleiman replied:
good points. rereading the spec:

>The initial-default is set only when a field is added to an existing schema. The write-default is initially set to the same value as initial-default and can be changed through schema evolution. If either default is not set for an optional field, then the default value is null for compatibility with older spec versions.

suggests that this isn't required at all, assuming we don't support required fields. so maybe the correct thing is to

- omit the `*-default` fields
- leave a TODO to the effect of "when/if we add support for required fields, 'initial-default' and 'write-default' should be written here" or something like that

oleiman replied:
done, as described above

andrwng commented on src/v/iceberg/update_schema_action.cc:
```diff
@@ -41,9 +41,13 @@ ss::future<action::action_outcome> update_schema_action::build_updates() && {
-    // the new fields.
-    new_schema_.assign_fresh_ids();
+    // NOTE: assumes that
+    //   - the table's last_column_id increases monotonically
+    //   - carry-over fields in a compat-checked schema already have
+    //     IDs assigned
+    //   - those carry-over IDs are all strictly <= table.last
```
Not suggesting we change even if it isn't the case, but just curious if these are the same invariants Iceberg's Java impl has?

oleiman replied:
which part? (2) and (3) are implementation details viz. 2) we update incoming fields w/ existing IDs and 3) that the last column ID of the table is correct with respect to the columns therein

(1) is perhaps expressed here: 

https://github.com/apache/iceberg/blob/8e456aeeabd0a40b23864edadd622b45cb44572c/core/src/main/java/org/apache/iceberg/TableMetadata.java#L577-L581

Which matches the new logic in table_update_applier.

andrwng commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -227,3 +385,472 @@ def run_verifier(schema: str, mapping: str):
+
+
+class Schema(NamedTuple):
+    schema: str
+    example: Callable[[float], dict]
+    spark_rep: list[tuple]
+    trino_rep: list[tuple]
+    select: list[str]
```
nit: mind adding comments about what each are?

nit: also maybe name it something to the effect SchemaTestCase or SchemaExpectations or something, given it's not just a schema?

oleiman replied:
done

oleiman committed:
ice/compat: Remove unused headers from compatibility_utils.h

Header only modules are not subject to dependency checks by bazel, so these
were causing linker errors in a dependent test.

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/7d81009699d72ef16cfb90f893160c4bc5c4245f..17eee074a877bd9d5e34e390909234c633dad513) CR comments

rockwotj commented on src/v/datalake/record_translator.cc:
```diff
@@ -188,6 +189,11 @@ structured_data_translator::build_type(std::optional<resolved_type> val_type) {
     if (val_type.has_value()) {
         val_id = std::move(val_type->id);
         auto& struct_type = std::get<iceberg::struct_type>(val_type->type);
+        // TODO: could make this configurable at some point
```
Can we not leave TODOs here and instead give discourse as to *why* do this.

oleiman replied:
done

rockwotj commented on src/v/iceberg/datatypes_json.cc:
```diff
@@ -177,6 +177,9 @@ void rjson_serialize(iceberg::json_writer& w, const iceberg::nested_field& f) {
     w.Bool(bool(f.required));
     w.Key("type");
     rjson_serialize(w, f.type);
+    // TODO: when/if we add support for required fields, 'initial-default' and
+    // 'write-default' should be written here
+    // See  https://iceberg.apache.org/spec/#default-values for requirements
```
So this isn't forgotten can we throw if `f.required` is true?

Also can we use the form of `TODO(iceberg)` so they are more greppable.

oleiman replied:
yeah good call

oleiman replied:
done

oleiman replied:
>throw if f.required is true

ah, right the redpanda header struct and sub-fields are still required... my intuition is that we should leave them as such because they are not evaluated with respect to the record schema but I'm not sure.

i'm going to remove the exception from this PR and spin off a separate ticket to avoid blowing up the diff here

oleiman replied:
i might argue that raising the exception from here, in the JSON serde code, is too "low level" for the thing we're trying to flag. policy vs representation or something like that.

rockwotj replied:
yeah fair point I wish this was decouples but here we are, I am good with just a comment

rockwotj commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -227,3 +385,480 @@ def run_verifier(schema: str, mapping: str):
+    total: int = 0
+
+
+class SchemaEvolutionE2ETests(RedpandaTest):
```
There is a *lot* here for these tests. The tests themselves are great, but I'd love to be able to describe them more concisely.

rockwotj replied:
I would love to have tests that have a higher level description and are translated into the specifics. As an example if we had something like this:

```
COMPATIBLE_CHANGES = {
  "add_column": EvolutionTest(
    type = "avro",
    initial_schema = {
    "type": "record",
    "name": "VerifierRecord",
    "fields": [
        {"name":"foo", "type": "string"},
        {
            "name": "bar",
            "type": {
                "type": "record",
                "name": "bar",
                "fields": [
                    {"name": "qux", "type": "int"},
                    {"name": "thud", "type": "bool"}
                ]
            }
        }
    ]},
    next_schema = {
    "type": "record",
    "name": "VerifierRecord",
    "fields": [
        {"name":"foo", "type": "string"},
        {
            "name": "bar",
            "type": {
                "type": "record",
                "name": "bar",
                "fields": [
                    {"name": "qux", "type": "int"},
                    {"name": "thud", "type": "bool"}
                ]
            }
        },
        {"name": "baz", "type": "string}
    ]},
    initial_table = [
     ("foo", "string"),
     ("bar", "struct", (
       "qux", "int",
       "thud", bool",
     )),
    ],
    next_table = [
     ("foo", "string"),
     ("bar", "struct", (
       "qux", "int",
       "thud", bool",
     )),
     ("baz", "string")
    ],
  ),
  "add_nested": EvolutionTest(...),
  ...
}

INCOMPATIBLE_CHANGES = {
  "illegal_promotion": EvolutionTest(...),
  ...,
}
```

Although even that is probably a lot... I wonder if we can do something like:

```
  "add_column": EvolutionTest(
    type = "avro",
    initial_schema = {
    "type": "record",
    "name": "VerifierRecord",
    "fields": [
        {"name":"foo", "type": "string"},
        {
            "name": "bar",
            "type": {
                "type": "record",
                "name": "bar",
                "fields": [
                    {"name": "qux", "type": "int"},
                    {"name": "thud", "type": "bool"}
                ]
            }
        }
    ]},
    next_schema = lambda og_schema: og_schema.fields.append({"name": "baz", "type": "string"}),
    initial_table = [
     ("foo", "string"),
     ("bar", "struct", (
       "qux", "int",
       "thud", bool",
     )),
    ],
    next_table = lambda og_schema: og_schmea.append(("baz", "string")),
  ),
```

Anyways not blocking, but I think that would help from making this file sprawl with all the cases here.

oleiman replied:
yeah I rather like the idea of representing the schema as a python object then evolution as diffs. will be less text, surely.

as far as the test functions themselves, I didn't have much luck distilling the inputs to something more general than "some schemas and some fields", since tests have pretty distinct semantics. 

there's a lot of repeated code though. could pull out another context manager for the shared stuff maybe. I'll take another run at it tonight.

oleiman replied:
🐍 🕳️

rockwotj commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -7,10 +7,17 @@
+import pyhive
 
+from confluent_kafka import avro
+from confluent_kafka.avro import AvroProducer
```
Are we planning on having protobuf support here too?

oleiman replied:
yeah should do. the code we're testing is downstream of schema translation, but there's no harm in more verification. these tests aren't very expensive to run.

may kick it to a follow-up to unblock #24955 though

rockwotj replied:
Totally 👍

rockwotj reviewed (commented):
Looks great! Thanks, I have a few comments and questions, mostly around tests, but that's not necessarily blocking...

oleiman committed:
dt/dl: Remove obsolete schema_evolution_test

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/17eee074a877bd9d5e34e390909234c633dad513..97b3e10e6ef665f6ca2daf41b21d55b869ad2d13) reworked ducktape tests, removed an old one, couple minor CR comments

oleiman commented:
/dt

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/b9ca4b4c5bee12aac14b6ca4b31d0c69a80839d3..605246c36418da4440cf43b1e184f0e9f7b46d35) rebase dev to fix conflict

rockwotj commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -20,210 +29,560 @@
+            ],
+        ),
+    ),
+    "drop_column":
```
I still think this is too aggressive of behavior, I rarely see people dropping columns and old data in analytical datasets, but that's not really a discussion for this PR.

oleiman replied:
Out of curiosity, do you mean

1. People won't use this - they will never (or rarely) drop fields from their record schemas
2. People will use this - they will drop fields from record schemas, but they will expect (or want) different behavior with respect to table schema

rockwotj replied:
I think #2 and ideally long term we gracefully having the user manually drop the column

oleiman replied:
Interesting, yeah. Variations on the theme of strict vs flexible interpretation of record schema. 

I don't have strong intuition one way or the other. Are we collecting open questions like this somewhere? If not I can start a fresh doc

rockwotj commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -20,210 +29,560 @@
+    next_schema: AvroSchema
+
+
+LEGAL_TEST_CASES = {
```
This is so great thank you I think its' way easier to grok when everything is colocated together ❤️

rockwotj commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -20,210 +29,560 @@
+            table = [(t[0], t[1]) for t in table[1:]]
+
+        assert table == expected.table(query_engine), \
+            str(table)
```
Can we add a little more color here about what was expected?

oleiman replied:
Protobuf diff is close behind, i'll be sure to improve these assert messages in there.

rockwotj commented on tests/rptest/tests/datalake/schema_evolution_test.py:
```diff
@@ -20,210 +29,560 @@
+            assert len(select_out) == count*3, \
+                f"Expected {count*3} rows, got {select_out}"
+
+            assert all(r[1] is None for r in select_out[:count * 2])
+            assert all(r[1] is not None for r in select_out[count * 2:])
```
Please add assertion messages

rockwotj reviewed (approved):
![giphy](https://github.com/user-attachments/assets/c3e0ae72-a491-4693-b27e-013a76882a5e)
